### PR TITLE
Implemented Backbone authentication

### PR DESCRIPTION
**Pull request by Trinovantes**
Going to work on the ~~CSS and~~ tests in future PRs

### CODE REVIEW

**Comment by Trinovantes** [comment]
Actually I'm going to commit css changes to this PR tomorrow. Can you guys just review the JS code for now?

**Comment by Trinovantes** [comment]
screenshot from 2015-10-24 21 12 34
screenshot from 2015-10-24 21 12 45

File: src-backbone/app/js/models/session.js [hunk 1]
```diff
@@ -0,0 +1,107 @@
+var AUTH_TOKEN_KEY = 'auth_token';
+var STORAGE_KEY = AUTH_TOKEN_KEY;
+
+var Config = require('utils/config');
+var Helpers = require('utils/helpers');
+
+var SessionModel = Backbone.Model.extend({
+
+    initialize: function() {
+        var me = this;
```
Comment (Ommy) [inline src-backbone/app/js/models/session.js]: I don't like that name 😛

File: src-backbone/app/js/utils/helpers.js [hunk 2]
```diff
@@ -0,0 +1,35 @@
+var Config = require('utils/config');
+
+module.exports = {
+
+    goto_default_logged_in: function() {
```
Comment (Ommy) [inline src-backbone/app/js/utils/helpers.js]: Why are method names here snake case? I'm not sure if it's different for helper methods but isn't convention to use camel case?

**Comment by Ommy** [comment]
LGTM once my comments have been addressed.

File: src-backbone/app/js/models/session.js (hunk 1 above)
Comment (Trinovantes) [inline src-backbone/app/js/models/session.js]: what about `dis`

File: src-backbone/app/js/utils/helpers.js (hunk 2 above)
Comment (Trinovantes) [inline src-backbone/app/js/utils/helpers.js]: I think most JS code bases use camel case but I used snake case to keep it consistent with our backend. Remember a few months ago when Django returned keys in snake case and Ember expected camel case?

File: src-backbone/app/js/models/session.js (hunk 1 above)
Comment (ConnorCimowsky) [inline src-backbone/app/js/models/session.js]: `self` please

File: src-backbone/app/index.html [hunk 3]
```diff
@@ -1,16 +1,22 @@
...
+    <meta http-equiv="X-UA-Compatible" content="IE=edge,chrome=1">
+    <meta name="viewport" content="width=device-width">
 
-  <link rel="stylesheet" href="assets/app.css">
-  <script src="assets/libs.js"></script>
-  <script src="assets/app.js"></script>
+    <link rel="shortcut icon" href="/img/favicon.ico" type="image/x-icon" />
+    <link rel="stylesheet" href="/css/libs.css">
+    <link rel="stylesheet" href="/css/app.css">
+
```
Comment (ConnorCimowsky) [inline src-backbone/app/index.html]: whitespace :sunglasses:

File: src-backbone/app/js/behaviors/authentication_form.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+        'submit @ui.form': 'auth',
+    },
+
+    show_errors_in_form: function($form, errors) {
+        $form.find('.alert').remove();
+        $form.find('.has-error').removeClass('has-error');
+        $form.find('span.help-block').remove();
+
+        Object.keys(errors).forEach(function(key) {
+            if (key === '__all__') {
```
Comment (ConnorCimowsky) [inline src-backbone/app/js/behaviors/authentication_form.js]: What does `__all__` mean?

File: src-backbone/app/js/behaviors/authentication_form.js [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+                    $parents.append('<span class="help-block">' + errors[key] + '</span>');
+                }
+            }
+        });
+    },
+
+    auth: function(event) {
+        event.preventDefault();
+
+        var form_view = this;
```
Comment (ConnorCimowsky) [inline src-backbone/app/js/behaviors/authentication_form.js]: camel case?

File: src-backbone/app/js/behaviors/authentication_form.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+
+    ui: {
+        form: 'form'
+    },
+
+    events: {
+        'submit @ui.form': 'auth',
+    },
+
+    show_errors_in_form: function($form, errors) {
```
Comment (ConnorCimowsky) [inline src-backbone/app/js/behaviors/authentication_form.js]: camel case?

**Comment by ConnorCimowsky** [comment]
I'll stop commenting on each instance, but I'm generally confused about what should or shouldn't be camel cased vs. snake cased. ¯\_(ツ)_/¯

File: src-backbone/app/js/models/session.js [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+        $.ajaxPrefilter(function(options, originalOptions, jqXHR) {
+            options.url = Config.API_BASE + options.url;
+            if (typeof me.get(AUTH_TOKEN_KEY) !== 'undefined') {
```
Comment (ConnorCimowsky) [inline src-backbone/app/js/models/session.js]: Is there a cleaner way of doing this check? I'm guessing `if (me.get(AUTH_TOKEN_KEY))` doesn't work?

**Comment by ConnorCimowsky** [comment]
re: camel case, I just re-read your comment, and I'm fine with whatever as long as we're consistent

File: src-backbone/app/js/behaviors/authentication_form.js (hunk 4 above)
Comment (Trinovantes) [inline src-backbone/app/js/behaviors/authentication_form.js]: That's a default key that Django forms return when there's a validation error for the entire form. e.g. Invalid username/password

File: src-backbone/app/index.html (hunk 3 above)
Comment (Trinovantes) [inline src-backbone/app/index.html]: Delete the new line?

Comment (ConnorCimowsky) [inline src-backbone/app/index.html]: I just meant there are 4 spaces on an empty line :stuck_out_tongue:

File: src-backbone/app/js/models/session.js (hunk 7 above)
Comment (Trinovantes) [inline src-backbone/app/js/models/session.js]: Good catch. Backbone models have a `has` method that does this http://backbonejs.org/#Model-has

File: src-backbone/app/index.html (hunk 3 above)
Comment (Trinovantes) [inline src-backbone/app/index.html]: Ah gotcha

**Comment by ConnorCimowsky** [comment]
I don't fully understand everything in here, but LGTM

File: src-backbone/app/js/behaviors/authentication_form.js (hunk 5 above)
Comment (Ommy) [inline src-backbone/app/js/behaviors/authentication_form.js]: This appears to be correct, at least according to: http://www.j-io.org/Javascript-Naming_Conventions/#variables

"Variables with multiple words should always use an underscore between words."

**Pull request merged**